### PR TITLE
[management] support the remainder of governance tasks

### PR DESCRIPTION
**Pull request by davidiw**
The purpose was to stress test the code and identify any missing pieces.

Features added:
* set-validator-operator
* create-validator
* create-validator-operator

Issues:
* set-validator-config fails if the validator is out of the set -- fixed by checking to see if they are in the set

Verified via some ugliness in libra-swarm:
* start
* grab one of the storage files
* make a config (see bottom, make sure it points to json server and proper storage)
* hack code to print out public keys for owner and operator keys in current namespace
* create a validator account and manually set `libra_root/owner` in local storage
* create an operator account and manually set `libra_root/operator` in local storage
* run set-validator-operator
* run set-validator-config
* run add-validator
* confirm validator-set (note this also kills libra since there are 2 validators but only 1 present)

```
chain_id: 4
json_server: "http://127.0.0.1:51928"
shared_backend:
    type: "on_disk_storage"
    path: "/Users/davidiw/libra/config/management/operational/storage"
    namespace: "libra_root"
validator_backend:
    type: "on_disk_storage"
    path: "/Users/davidiw/libra/config/management/operational/storage"
    namespace: "libra_root"
```

@mgorven, we'll be using this along with the genesis tooling to get things setup...

### CODE REVIEW

File: config/management/operational/src/governance.rs [hunk 1]
```diff
@@ -2,11 +2,104 @@
...
+impl CreateValidator {
+    pub fn execute(self) -> Result<(TransactionContext, AccountAddress), Error> {
+        let config = self.input.config()?;
+        let key = self.input.key()?;
+        let client = JsonRpcClientWrapper::new(config.json_server.clone());
+
+        let seq_num = client.sequence_number(libra_root_address())?;
+        let auth_key = AuthenticationKey::ed25519(&key);
+        let account_address = auth_key.derived_address();
+        let script = transaction_builder::encode_create_validator_account_script(
```
Comment (JoshLind) [inline config/management/operational/src/governance.rs:64]: This flow diverges from the one in genesis (where we also allow auth keys to be none in the case that the owner doesn't provide a key..). Do we want to support this functionality too here, or just leave it as is and use a throw away key?

File: config/management/operational/src/governance.rs [hunk 2]
```diff
@@ -2,11 +2,104 @@
...
+}
+
+#[derive(Debug, StructOpt)]
+pub struct CreateValidatorOperator {
+    #[structopt(flatten)]
+    input: CreateAccount,
+}
+
+impl CreateValidatorOperator {
+    pub fn execute(self) -> Result<(TransactionContext, AccountAddress), Error> {
```
Comment (JoshLind) [inline config/management/operational/src/governance.rs:83]: nit: it seems the only difference between this execute() call and the one for the CreateValidator struct is the script type (encode_create_validator_operator_account_script vs encode_create_validator_account_script). Ideally we should refactor into a common helper (e.g., on CreateAccount) to avoid duplication?

File: config/management/operational/src/governance.rs [hunk 3]
```diff
@@ -80,21 +164,27 @@ impl RemoveValidator {
...
+
+fn build_and_submit_libra_root_transaction(
+    config: &Config,
+    seq_num: u64,
+    script: Script,
+) -> Result<TransactionContext, Error> {
+    let txn = build_raw_transaction(config.chain_id, libra_root_address(), seq_num, script);
+
+    let mut storage = config.validator_backend();
+    let signed_txn = storage.sign(LIBRA_ROOT_KEY, "remove-validator", txn)?;
```
Comment (JoshLind) [inline config/management/operational/src/governance.rs:186]: nit: the script name "remove-validator" needs to be fixed.. Perhaps pass it directly into this function from the caller?

File: config/management/operational/src/owner.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            config.chain_id,
+            owner_address,
+            client.sequence_number(owner_address)?,
+            transaction_builder::encode_set_validator_operator_script(
+                self.name.as_bytes().to_vec(),
+                self.account_address,
+            ),
+        );
+
+        println!(
```
Comment (JoshLind) [inline config/management/operational/src/owner.rs:44]: nit: do we still want to keep these in here?

File: config/management/operational/src/validator_config.rs [hunk 5]
```diff
@@ -55,32 +55,45 @@ impl SetValidatorConfig {
...
+            "/ip4/127.0.0.1/tcp/6180".parse().unwrap()
         };
 
         let fullnode_address = if let Some(fullnode_address) = self.fullnode_address {
             fullnode_address
+        } else if let Some(vc) = &validator_config {
+            strip_address(&vc.fullnode_network_address)
         } else {
-            strip_address(&validator_config.fullnode_network_address)
+            "/ip4/127.0.0.1/tcp/6180".parse().unwrap()
```
Comment (gregnazario) [inline config/management/operational/src/validator_config.rs:89]: So, what happens when we do these loopback addresses?  For example, another peer tries to connect to this peer, and finds out that it's addresses are localhost, it would connect to itself, and then realize the peerId doesn't match?

File: config/management/operational/src/validator_config.rs [hunk 6]
```diff
@@ -55,32 +55,45 @@ impl SetValidatorConfig {
...
+            None
+        };
 
         let validator_address = if let Some(validator_address) = self.validator_address {
             validator_address
+        } else if let Some(vc) = &validator_config {
+            strip_address(&vc.validator_network_address)
         } else {
-            strip_address(&validator_config.validator_network_address)
+            "/ip4/127.0.0.1/tcp/6180".parse().unwrap()
```
Comment (JoshLind) [inline config/management/operational/src/validator_config.rs:81]: Can you document why we want these to be the defaults? Or is this related to the debugging you did locally?

File: config/management/operational/src/governance.rs (hunk 3 above)
Comment (gregnazario) [inline config/management/operational/src/governance.rs:186]: Do you want to sign it with "remove-validator" when you also add a validator?

**Review by JoshLind — APPROVED** [review]
Thanks @davidiw. 😄  I'll approve this now so that you can move fast for the drills if need be, but please see my comments.

File: config/management/operational/src/owner.rs (hunk 4 above)
Comment (davidiw) [reply config/management/operational/src/owner.rs:44]: hahahaha oops

File: config/management/operational/src/governance.rs (hunk 3 above)
Comment (davidiw) [reply config/management/operational/src/governance.rs:186]: oh yeah, I saw this and forgot to comeback .... problem of working too late at night...

File: config/management/operational/src/governance.rs (hunk 2 above)
Comment (davidiw) [reply config/management/operational/src/governance.rs:83]: fair enough

File: config/management/operational/src/governance.rs (hunk 1 above)
Comment (davidiw) [reply config/management/operational/src/governance.rs:64]: The problem is what is exposed via the public interfaces versus the admin scripts. This is the best we can do with public interfaces. It most certainly does not match what genesis will do whereas other systems will. We can follow up later to explore admin scripts to generate the account, but definitely not critical for e2e tests.

File: config/management/operational/src/governance.rs (hunk 3 above)
Comment (davidiw) [reply config/management/operational/src/governance.rs:186]: that's just a message for errors, I need to pass in a string here...

File: config/management/operational/src/validator_config.rs (hunk 5 above)
Comment (davidiw) [reply config/management/operational/src/validator_config.rs:89]: the better behavior is probably to raise an error... I'll update...

File: config/management/operational/src/validator_config.rs (hunk 6 above)
Comment (davidiw) [reply config/management/operational/src/validator_config.rs:81]: Did you and @gregnazario work together on reviewing this, rofl... I'll raise an error.

File: config/management/operational/src/governance.rs (hunk 1 above)
Comment (JoshLind) [reply config/management/operational/src/governance.rs:64]: Sure, that makes sense! Just wanted to bring it up to make sure we're aware of the divergence 😄
